src/libcrun/linux.c

```diff
@@ -5713,3 +5719,38 @@ libcrun_update_intel_rdt (const char *ctr_name, libcrun_container_t *container,
 
   return resctl_update (name, l3_cache_schema, mem_bw_schema, err);
 }
+
+/* Change the current directory and make sure the current working
+   directory, once set, is accessible from the current mount
+   namespace.  This check prevents container-escape issues like
+   CVE-2024-21626.
+   The current working directory cannot be longer than PATH_MAX.
+*/
+int
+libcrun_safe_chdir (const char *path)
+{
+  cleanup_free char *buffer = NULL;
+  int ret;
+
+  ret = chdir (path);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  buffer = xmalloc (PATH_MAX);
+  ret = syscall_getcwd (buffer, PATH_MAX);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  /* Enforce that the returned path is an absolute path.
+     The kernel prepends the string "(unreachable)" to the path
+     when it is not reachable from the current mount namespace.
+     The following check also detects the "unreachable" case.
+  */
+  if (ret == 0 || buffer[0] != '/')
+    {
+      (void) chdir ("/");
+      errno = ENOENT;
```

Comment:
Should we print a big error message if this happens, that some container just tried to fool us?

Reply:
improved the error message and pushed a new version